// @atlaskit/avatar/dist/es/components/Presence.js
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import _PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import { Inner, Outer } from '../styled/Icon';
import getPresenceSVG from '../helpers/getPresenceSVG';

var Presence = function (_PureComponent) {
  _inherits(Presence, _PureComponent);

  function Presence() {
    _classCallCheck(this, Presence);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  // eslint-disable-line react/sort-comp

  Presence.prototype.render = function render() {
    var _props = this.props,
        borderColor = _props.borderColor,
        children = _props.children,
        presence = _props.presence,
        size = _props.size;


    return React.createElement(
      Outer,
      { size: size, bgColor: borderColor },
      React.createElement(
        Inner,
        null,
        children || presence && getPresenceSVG(presence)
      )
    );
  };

  return Presence;
}(PureComponent);

Presence.propTypes = {
  /** Used to override the default border color of the presence indicator.
  Accepts any color argument that the border-color CSS property accepts. */
  borderColor: _PropTypes.oneOfType([_PropTypes.string, _PropTypes.func]) /** Content to use as a custom presence indicator (usually not required if
                                                                          consuming Presence separate to Avatar). */
  ,
  children: _PropTypes.any /** Content to use as a custom presence indicator (usually not required if
                           consuming Presence separate to Avatar). */
  ,
  presence: _PropTypes.oneOfType([_PropTypes.oneOf(['online', 'busy', 'offline']), _PropTypes.any]) /** Defines the size of the presence. */
  ,
  size: _PropTypes.oneOf(['xsmall', 'small', 'medium', 'large', 'xlarge', 'xxlarge'])
};
export default Presence;
