// @atlaskit/avatar/dist/es/components/Status.js
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import _PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import { Outer } from '../styled/Icon';
import getStatusSVG from '../helpers/getStatusSVG';

var Status = function (_PureComponent) {
  _inherits(Status, _PureComponent);

  function Status() {
    _classCallCheck(this, Status);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  // eslint-disable-line react/sort-comp

  Status.prototype.render = function render() {
    var _props = this.props,
        borderColor = _props.borderColor,
        children = _props.children,
        status = _props.status,
        size = _props.size;


    return React.createElement(
      Outer,
      { size: size, bgColor: borderColor },
      children || status && getStatusSVG(status)
    );
  };

  return Status;
}(PureComponent);

Status.propTypes = {
  /** Used to override the default border color of the status indicator.
  Accepts any color argument that the border-color CSS property accepts. */
  borderColor: _PropTypes.oneOfType([_PropTypes.string, _PropTypes.func]) /** Content to use as a custom status indicator (usually not required if
                                                                          consuming Status separate to Avatar). */
  ,
  children: _PropTypes.any /** Content to use as a custom status indicator (usually not required if
                           consuming Status separate to Avatar). */
  ,
  status: _PropTypes.oneOf(['approved', 'declined', 'locked']) /** Defines the size of the status. */
  ,
  size: _PropTypes.oneOf(['xsmall', 'small', 'medium', 'large', 'xlarge', 'xxlarge'])
};
export default Status;
