// @atlaskit/multi-select/dist/es/components/Footer.js
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import { Item } from '@atlaskit/droplist';
import FooterDiv from '../styled/Footer';

var Footer = function (_PureComponent) {
  _inherits(Footer, _PureComponent);

  function Footer() {
    _classCallCheck(this, Footer);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  Footer.prototype.render = function render() {
    var _props = this.props,
        appearance = _props.appearance,
        children = _props.children,
        elemBefore = _props.elemBefore,
        isFocused = _props.isFocused,
        onClick = _props.onClick,
        shouldHideSeparator = _props.shouldHideSeparator;


    return React.createElement(
      FooterDiv,
      { onClick: onClick, shouldHideSeparator: shouldHideSeparator },
      React.createElement(
        Item,
        {
          appearance: appearance,
          elemBefore: elemBefore,
          isFocused: isFocused,
          type: 'option'
        },
        children
      )
    );
  };

  return Footer;
}(PureComponent);

Footer.propTypes = {
  appearance: PropTypes.oneOf(['default', 'primary']),
  children: PropTypes.node,
  elemBefore: PropTypes.node,
  isFocused: PropTypes.bool,
  onClick: PropTypes.func,
  shouldHideSeparator: PropTypes.bool
};
export default Footer;
