// @atlaskit/flag/dist/es/components/FlagActions.js
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import _PropTypes from 'prop-types';
import React, { PureComponent } from 'react';
import Container, { Action, Button } from '../styled/Actions';

import { DEFAULT_APPEARANCE } from './Flag';

var FlagActions = function (_PureComponent) {
  _inherits(FlagActions, _PureComponent);

  function FlagActions() {
    _classCallCheck(this, FlagActions);

    return _possibleConstructorReturn(this, _PureComponent.apply(this, arguments));
  }

  // eslint-disable-line react/sort-comp
  FlagActions.prototype.render = function render() {
    var _props = this.props,
        actions = _props.actions,
        appearance = _props.appearance;

    var isBold = appearance !== DEFAULT_APPEARANCE;

    if (!actions.length) return null;

    var items = actions.map(function (action, index) {
      return React.createElement(
        Action,
        { key: index, hasDivider: !!index, useMidDot: !isBold },
        React.createElement(
          Button,
          { onClick: action.onClick, appearance: appearance },
          action.content
        )
      );
    });

    return React.createElement(
      Container,
      null,
      items
    );
  };

  return FlagActions;
}(PureComponent);

FlagActions.propTypes = {
  appearance: _PropTypes.oneOf(['error', 'info', 'normal', 'success', 'warning']).isRequired,
  actions: _PropTypes.arrayOf(_PropTypes.shape({
    content: _PropTypes.any.isRequired,
    onClick: _PropTypes.func.isRequired
  })).isRequired
};
FlagActions.defaultProps = {
  appearance: DEFAULT_APPEARANCE,
  actions: []
};
export default FlagActions;
