// @atlaskit/field-base/dist/es/styled/theme.js
import { colors, themed } from '@atlaskit/theme';

// The following are the name for color mappings in @atlaskit/themes
// The exports are the functions, not the objects, so could not be used here
var codeBlock = { light: colors.N20, dark: colors.DN80 };
var hover = { light: colors.N20, dark: colors.DN30 };
var disabled = { light: colors.N20, dark: colors.DN20 };
var yellow = { light: colors.Y300, dark: colors.Y300 };
// linkOutline has been altered based on the colors in the text-field spec
var linkOutline = { light: colors.B100, dark: colors.B75 };
// The following do not yet have a darkmode 'map': N20A, N10

export var getBackgroundColor = themed('appearance', {
  standard: { light: colors.N10, dark: colors.DN10 },
  disabled: disabled,
  invalid: { light: colors.N10, dark: colors.DN10 },
  subtle: { light: 'transparent', dark: 'transparent' },
  none: { light: 'transparent', dark: 'transparent' }
});

export var getBackgroundColorFocus = themed('appearance', {
  standard: { light: colors.N0, dark: colors.DN10 },
  disabled: disabled,
  invalid: { light: colors.N0, dark: colors.DN10 },
  subtle: { light: colors.N0, dark: colors.DN10 },
  none: { light: 'transparent', dark: 'transparent' }
});

export var getBackgroundColorHover = themed('appearance', {
  standard: hover,
  disabled: disabled,
  invalid: hover,
  subtle: hover,
  none: { light: 'transparent', dark: 'transparent' }
});

export var getBorderColor = themed('appearance', {
  standard: codeBlock,
  disabled: { light: colors.N20A, dark: colors.DN20A },
  invalid: yellow,
  subtle: { light: 'transparent', dark: 'transparent' },
  none: { light: 'transparent', dark: 'transparent' }
});

export var getBorderColorFocus = themed('appearance', {
  standard: linkOutline,
  disabled: { light: colors.N20A, dark: colors.DN20A },
  invalid: linkOutline,
  subtle: linkOutline,
  none: { light: 'transparent', dark: 'transparent' }
});

export var getBorderColorHover = themed('appearance', {
  standard: codeBlock,
  disabled: { light: colors.N20A, dark: colors.DN20A },
  invalid: yellow,
  subtle: codeBlock,
  none: { light: 'transparent', dark: 'transparent' }
});
