// @atlaskit/flag/dist/es/styled/Flag.js
import _taggedTemplateLiteral from 'babel-runtime/helpers/taggedTemplateLiteral';

var _templateObject = _taggedTemplateLiteral(['\n  background-color: ', ';\n  border-radius: ', 'px;\n  box-sizing: border-box;\n  box-shadow: ', ';\n  color: ', ';\n  display: flex;\n  padding: ', 'px;\n  transition: background-color 200ms;\n  width: 100%;\n  z-index: 600;\n\n  &:focus {\n    outline: none;\n    box-shadow: 0px 0px 0px 2px ', ';\n  }\n'], ['\n  background-color: ', ';\n  border-radius: ', 'px;\n  box-sizing: border-box;\n  box-shadow: ', ';\n  color: ', ';\n  display: flex;\n  padding: ', 'px;\n  transition: background-color 200ms;\n  width: 100%;\n  z-index: 600;\n\n  &:focus {\n    outline: none;\n    box-shadow: 0px 0px 0px 2px ', ';\n  }\n']),
    _templateObject2 = _taggedTemplateLiteral(['\n  display: flex;\n'], ['\n  display: flex;\n']),
    _templateObject3 = _taggedTemplateLiteral(['\n  color: ', ';\n  font-weight: 600;\n  flex: 1;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  white-space: nowrap;\n'], ['\n  color: ', ';\n  font-weight: 600;\n  flex: 1;\n  overflow: hidden;\n  text-overflow: ellipsis;\n  white-space: nowrap;\n']),
    _templateObject4 = _taggedTemplateLiteral(['\n  appearance: none;\n  background: none;\n  border: none;\n  border-radius: ', 'px;\n  color: ', ';\n  cursor: pointer;\n  display: flex;\n  margin-left: ', 'px;\n  padding: 0;\n  white-space: nowrap;\n\n  &:focus {\n    outline: none;\n    box-shadow: 0px 0px 0px 2px ', ';\n  }\n'], ['\n  appearance: none;\n  background: none;\n  border: none;\n  border-radius: ', 'px;\n  color: ', ';\n  cursor: pointer;\n  display: flex;\n  margin-left: ', 'px;\n  padding: 0;\n  white-space: nowrap;\n\n  &:focus {\n    outline: none;\n    box-shadow: 0px 0px 0px 2px ', ';\n  }\n']),
    _templateObject5 = _taggedTemplateLiteral(['\n  display: flex;\n  flex: 1 1 100%;\n  flex-direction: column;\n  justify-content: center;\n  min-width: 0;\n'], ['\n  display: flex;\n  flex: 1 1 100%;\n  flex-direction: column;\n  justify-content: center;\n  min-width: 0;\n']),
    _templateObject6 = _taggedTemplateLiteral(['\n  max-height: ', 'px;\n  opacity: ', ';\n  overflow: ', ';\n  transition: max-height 0.3s, opacity 0.3s;\n'], ['\n  max-height: ', 'px;\n  opacity: ', ';\n  overflow: ', ';\n  transition: max-height 0.3s, opacity 0.3s;\n']),
    _templateObject7 = _taggedTemplateLiteral(['\n  color: ', ';\n  margin-top: ', 'px;\n'], ['\n  color: ', ';\n  margin-top: ', 'px;\n']),
    _templateObject8 = _taggedTemplateLiteral(['\n  align-items: flex-start;\n  display: inline-flex;\n  flex: 0 0 auto;\n  flex-direction: column;\n  width: ', 'px;\n'], ['\n  align-items: flex-start;\n  display: inline-flex;\n  flex: 0 0 auto;\n  flex-direction: column;\n  width: ', 'px;\n']);

import styled from 'styled-components';

import { borderRadius, gridSize, math } from '@atlaskit/theme';

import { flagBackgroundColor, flagBorderColor, flagTextColor, flagShadowColor, flagFocusRingColor } from '../theme';

var getBoxShadow = function getBoxShadow(props) {
  var borderColor = flagBorderColor(props);
  var shadowColor = flagShadowColor(props);

  var border = borderColor && '0 0 1px ' + borderColor;
  var shadow = '0 20px 32px -8px ' + shadowColor;

  return [border, shadow].filter(function (p) {
    return p;
  }).join(',');
};

export default styled.div(_templateObject, flagBackgroundColor, borderRadius, getBoxShadow, flagTextColor, math.multiply(gridSize, 2), flagFocusRingColor);

// Header
export var Header = styled.div(_templateObject2);

export var Title = styled.span(_templateObject3, flagTextColor);

export var DismissButton = styled.button(_templateObject4, borderRadius, flagTextColor, gridSize, flagFocusRingColor);

// Content
export var Content = styled.div(_templateObject5);

export var Expander = styled.div(_templateObject6, function (_ref) {
  var isExpanded = _ref.isExpanded;
  return isExpanded ? 150 : 0;
}, function (_ref2) {
  var isExpanded = _ref2.isExpanded;
  return isExpanded ? 1 : 0;
}, function (_ref3) {
  var isExpanded = _ref3.isExpanded;
  return isExpanded ? 'visible' : 'hidden';
});

export var Description = styled.div(_templateObject7, flagTextColor, gridSize);

export var Icon = styled.div(_templateObject8, math.multiply(gridSize, 4));
