// @atlaskit/tag/dist/cjs/Chrome/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styled = require('./styled');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Chrome = function (_PureComponent) {
  (0, _inherits3.default)(Chrome, _PureComponent);

  function Chrome() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, Chrome);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = Chrome.__proto__ || (0, _getPrototypeOf2.default)(Chrome)).call.apply(_ref, [this].concat(args))), _this), _this.handleKeyPress = function (e) {
      var spacebarOrEnter = e.key === ' ' || e.key === 'Enter';

      if (_this.chromeRef && spacebarOrEnter) {
        var link = _this.chromeRef.querySelector('a');
        if (link) link.click();
      }
    }, _this.handleBlur = function () {
      _this.props.onFocusChange(false);
    }, _this.handleFocus = function (e) {
      if (e.target === _this.chromeRef) _this.props.onFocusChange(true);
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(Chrome, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          children = _props.children,
          isLink = _props.isLink,
          isRemovable = _props.isRemovable,
          isRemoved = _props.isRemoved,
          isRemoving = _props.isRemoving,
          isRounded = _props.isRounded,
          markedForRemoval = _props.markedForRemoval,
          color = _props.color;


      var props = {
        innerRef: function innerRef(r) {
          _this2.chromeRef = r;
        },
        isRemovable: isRemovable,
        isRemoved: isRemoved,
        isRemoving: isRemoving,
        isRounded: isRounded,
        markedForRemoval: markedForRemoval,
        onBlur: this.handleBlur,
        onFocus: this.handleFocus,
        onKeyPress: this.handleKeyPress,
        tabIndex: -1,
        color: color,
        role: ''
      };

      if (isLink) {
        props.role = 'link';
        props.tabIndex = 0;
      }

      return _react2.default.createElement(
        _styled.Span,
        props,
        children
      );
    }
  }]);
  return Chrome;
}(_react.PureComponent);

exports.default = Chrome;
