// @atlaskit/tag/dist/cjs/RemoveButton/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _cross = require('@atlaskit/icon/glyph/cross');

var _cross2 = _interopRequireDefault(_cross);

var _styled = require('./styled');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var RemoveButton = function (_PureComponent) {
  (0, _inherits3.default)(RemoveButton, _PureComponent);

  function RemoveButton() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, RemoveButton);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = RemoveButton.__proto__ || (0, _getPrototypeOf2.default)(RemoveButton)).call.apply(_ref, [this].concat(args))), _this), _this.onKeyPress = function (e) {
      var spacebarOrEnter = e.key === ' ' || e.key === 'Enter';

      if (spacebarOrEnter) {
        e.stopPropagation();

        if (_this.props.onRemoveAction) {
          _this.props.onRemoveAction();
        }
      }
    }, _this.onMouseOver = function () {
      if (_this.props.onHoverChange) _this.props.onHoverChange(true);
    }, _this.onMouseOut = function () {
      if (_this.props.onHoverChange) _this.props.onHoverChange(false);
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(RemoveButton, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          isRounded = _props.isRounded,
          onRemoveAction = _props.onRemoveAction,
          removeText = _props.removeText;


      return _react2.default.createElement(
        _styled.Button,
        {
          'aria-label': removeText,
          isRounded: isRounded,
          onClick: onRemoveAction,
          onKeyPress: this.onKeyPress,
          onMouseOut: this.onMouseOut,
          onMouseOver: this.onMouseOver,
          type: 'button'
        },
        _react2.default.createElement(_cross2.default, { label: removeText, size: 'small' })
      );
    }
  }]);
  return RemoveButton;
}(_react.PureComponent);

exports.default = RemoveButton;
