// @atlaskit/dropdown-menu/dist/es/components/group/DropdownItemGroup.js
import _classCallCheck from 'babel-runtime/helpers/classCallCheck';
import _possibleConstructorReturn from 'babel-runtime/helpers/possibleConstructorReturn';
import _inherits from 'babel-runtime/helpers/inherits';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ItemGroup } from '@atlaskit/item';

var DropdownItemGroup = function (_Component) {
  _inherits(DropdownItemGroup, _Component);

  function DropdownItemGroup() {
    _classCallCheck(this, DropdownItemGroup);

    return _possibleConstructorReturn(this, _Component.apply(this, arguments));
  }

  DropdownItemGroup.prototype.render = function render() {
    var _props = this.props,
        children = _props.children,
        elemAfter = _props.elemAfter,
        title = _props.title;

    return React.createElement(
      ItemGroup,
      {
        elemAfter: elemAfter,
        title: title,
        role: 'menu'
      },
      children
    );
  };

  return DropdownItemGroup;
}(Component);

DropdownItemGroup.propTypes = {
  /** DropdownItems to be rendered inside the group.*/
  children: PropTypes.node,
  /** Optional heading text to be shown above the items. */
  title: PropTypes.string,
  /** Content to be shown to the right of the title heading. Not shown if no title is set. */
  elemAfter: PropTypes.oneOfType([PropTypes.node, PropTypes.string])
};
export default DropdownItemGroup;
